packages/app/src/server/service/page.ts

```diff
@@ -2254,11 +2254,23 @@ class PageService {
   async normalizeParentByPageIds(pageIds: ObjectIdLike[], user, isRecursively: boolean): Promise<void> {
     const Page = mongoose.model('Page') as unknown as PageModel;
 
+    const socket = this.crowi.socketIoService.getDefaultSocket();
+
+    socket.emit(SocketEventName.PageMigrationStarted);
+
     if (isRecursively) {
       const pages = await Page.findByIdsAndViewer(pageIds, user, null);
 
       // DO NOT await !!
-      this.normalizeParentRecursivelyByPages(pages, user);
+      (async() => {
+        try {
+          await this.normalizeParentRecursivelyByPages(pages, user);
+          socket.emit(SocketEventName.PageMigrationEnded);
+        }
+        catch {
+          socket.emit(SocketEventName.PageMigrationError);
+        }
+      })();
```

Comment:
`normalizeParentRecursivelyByPages` は複数のページである `pages` をとって、それらのページがコンバート可能かどうかを判断する処理をやってる。

```typescript:page.ts
let normalizablePages;
    let nonNormalizablePages;
    try {
      [normalizablePages, nonNormalizablePages] = await this.crowi.pageGrantService.separateNormalizableAndNotNormalizablePages(user, pagesToNormalize);
    }
    catch (err) {
      throw err;
    }

    if (normalizablePages.length === 0) {
      // socket.emit('normalizeParentRecursivelyByPages', { error: err.message }); TODO: use socket to tell user
      return;
    }

    if (nonNormalizablePages.length !== 0) {
      // TODO: iterate nonNormalizablePages and send socket error to client so that the user can know which path failed to migrate
      // socket.emit('normalizeParentRecursivelyByPages', { error: err.message }); TODO: use socket to tell user
    }
```

ここで emit すると、ただ emit するだけじゃなく、どのページがコンバートできなかったかなどの情報までクライアントに送ることができる。のでここでやるのがいいかな。

あとどんなふうにエラーを見せようと考えているか教えて欲しい。